frontend/src/metabase/query_builder/actions.js

```diff
@@ -554,26 +547,23 @@ export const loadMetadataForCard = createThunkAction(
       if (!card || !card.dataset_query) {
         return;
       }
-      const query = new Question(card, getMetadata(getState())).query();
-      if (query instanceof StructuredQuery) {
-        try {
-          const rootTableId = query.rootTableId();
-          if (rootTableId != null) {
-            await Promise.all([
-              dispatch(Tables.actions.fetchTableMetadata({ id: rootTableId })),
-              dispatch(Tables.actions.fetchForeignKeys({ id: rootTableId })),
-            ]);
-          }
-          await Promise.all(
-            query
-              .dependentTableIds()
-              .map(id => dispatch(Tables.actions.fetchMetadata({ id }))),
-          );
-        } catch (e) {
-          console.error("Error loading metadata for card", e);
-          throw e;
-        }
-      }
+      const loadedIds = [];
+      let newIds;
+      // NOTE: need to do this in a loop so that implicitly linked (i.e. fk->) tables are loaded
+      // we don't know about them until the initial table is loaded
+      do {
+        const metadata = getMetadata(getState());
+        const query = new Question(card, metadata).query();
+        newIds = _.difference(query.dependentTableIds(), loadedIds);
+        await Promise.all(
+          newIds
+            // don't try to load tables we don't have access to
+            // this assumes the list of tables has been loaded
+            .filter(id => metadata.table(id))
+            .map(id => dispatch(Tables.actions.fetchMetadata({ id }))),
+        );
+        loadedIds.push(...newIds);
+      } while (newIds.length > 0);
```

Comment:
Will this loop ever happen more than twice?

If it can repeat more than twice, maybe it makes sense to have cards include a list of all dependent tables when returned from the server. That would avoid waiting on multiple round trips.

Reply:
I don't believe so.

Another alternative I considered was something like changing `dependentTableIds` to `dependentMetadata` which returns something like:

```
[
  { type: "table", id: 123, foreignTables: true },
  { type: "table", id: 234 },
  { type: "table", id: 345 }
]
```

That seems a little cleaner to me. We could also have `loadMetadataForQuestion` in `metabase/redux/metadata` that calls `dependentMetadata` and loads everything necessary.